File: backend/src/metrology/mpe.ts
import { Decimal, exactDecimal, MetrologyDomainError } from '../core/decimal.js';
import { AccuracyClassEnum } from '../core/types.js';

export interface NawiMpeResult {
  factorInE: Decimal;
  mpeInMass: Decimal;
  intervalsM: Decimal;
  isInitial: boolean;
}

/**
 * Computes statutory Maximum Permissible Error (MPE) for Non-Automatic Weighing Instruments (NAWI)
 * conforming to OIML R 76-1 and Legal Metrology (General) Rules, 2011 (Seventh Schedule).
 *
 * @param load Applied nominal test load in mass units (e.g. kg or g)
 * @param scaleIntervalE Verification scale interval 'e' in same mass units
 * @param accuracyClass Accuracy Class (CLASS_I, CLASS_II, CLASS_III, CLASS_IIII)
 * @param isInitialVerification True for initial/after-repair verification, false for in-service re-verification
 */
export function calculateNawiMpe(
  load: Decimal | string | number,
  scaleIntervalE: Decimal | string | number,
  accuracyClass: AccuracyClassEnum | string,
  isInitialVerification = true
): NawiMpeResult {
  const L = exactDecimal(load);
  const e = exactDecimal(scaleIntervalE);

  if (L.isNegative()) {
    throw new MetrologyDomainError(`Nominal load cannot be negative: ${L.toString()}`);
  }
  if (e.isZero() || e.isNegative()) {
    throw new MetrologyDomainError(`Verification scale interval 'e' must be strictly positive: ${e.toString()}`);
  }

  // Calculate load in number of scale intervals: m = L / e
  const m = L.dividedBy(e);

  let initialFactor: Decimal;

  const normalizedClass = accuracyClass.toUpperCase();

  switch (normalizedClass) {
    case 'CLASS_I':
    case 'I':
      if (m.lessThanOrEqualTo(50000)) {
        initialFactor = exactDecimal('0.5');
      } else if (m.lessThanOrEqualTo(200000)) {
        initialFactor = exactDecimal('1.0');
      } else {
        initialFactor = exactDecimal('1.5');
      }
      break;

    case 'CLASS_II':
    case 'II':
      if (m.lessThanOrEqualTo(5000)) {
        initialFactor = exactDecimal('0.5');
      } else if (m.lessThanOrEqualTo(20000)) {
        initialFactor = exactDecimal('1.0');
      } else {
        initialFactor = exactDecimal('1.5');
      }
      break;

    case 'CLASS_III':
    case 'III':
      if (m.lessThanOrEqualTo(500)) {
        initialFactor = exactDecimal('0.5');
      } else if (m.lessThanOrEqualTo(2000)) {
        initialFactor = exactDecimal('1.0');
      } else {
        initialFactor = exactDecimal('1.5');
      }
      break;

    case 'CLASS_IIII':
    case 'IIII':
    case 'CLASS_IV':
    case 'IV':
      if (m.lessThanOrEqualTo(50)) {
        initialFactor = exactDecimal('0.5');
      } else if (m.lessThanOrEqualTo(200)) {
        initialFactor = exactDecimal('1.0');
      } else {
        initialFactor = exactDecimal('1.5');
      }
      break;

    default:
      throw new MetrologyDomainError(`Unsupported accuracy class for MPE calculation: ${accuracyClass}`);
  }

  // Periodic re-verification MPE is 2 * Initial MPE (Seventh Schedule / Section 24)
  const multiplier = isInitialVerification ? exactDecimal('1.0') : exactDecimal('2.0');
  const factorInE = initialFactor.times(multiplier);
  const mpeInMass = factorInE.times(e);

  return {
    factorInE,
    mpeInMass,
    intervalsM: m,
    isInitial: isInitialVerification,
  };
}

File: packages/verification-procedures/nawi/mpe.py
"""Stepped Maximum Permissible Error (MPE) calculation engine for NAWI.

Implements statutory MPE stepped functions under The Legal Metrology (General) Rules, 2011
(Seventh Schedule, Part II, Table 4) and OIML R 76-1 Table 6.
"""

from __future__ import annotations

from packages.measurement.decimal_math import ExactDecimal, exact_decimal
from packages.measurement.units import Quantity
from ..base import AccuracyClassEnum, VerificationTypeEnum


def get_nawi_mpe_factor_in_e(
    m_intervals: ExactDecimal,
    accuracy_class: AccuracyClassEnum,
    verification_type: VerificationTypeEnum,
) -> ExactDecimal:
    """Determine the stepped MPE factor in units of verification scale interval e.

    Args:
        m_intervals: Load expressed in scale intervals (m = L / e).
        accuracy_class: Metrological accuracy class (Class I, II, III, IIII).
        verification_type: INITIAL or RE_VERIFICATION.

    Returns:
        ExactDecimal factor (e.g. 0.5, 1.0, 1.5 for initial or 1.0, 2.0, 3.0 for in-service).
    """
    m = exact_decimal(m_intervals)
    if m < ExactDecimal("0"):
        m = ExactDecimal(abs(m))

    multiplier = ExactDecimal("2") if verification_type == VerificationTypeEnum.RE_VERIFICATION else ExactDecimal("1")

    if accuracy_class == AccuracyClassEnum.CLASS_III:
        # Class III stepped boundaries: 500e, 2000e, 10000e
        if m <= ExactDecimal("500"):
            base_factor = ExactDecimal("0.5")
        elif m <= ExactDecimal("2000"):
            base_factor = ExactDecimal("1.0")
        else:
            base_factor = ExactDecimal("1.5")

    elif accuracy_class == AccuracyClassEnum.CLASS_IIII:
        # Class IIII stepped boundaries: 50e, 200e, 1000e
        if m <= ExactDecimal("50"):
            base_factor = ExactDecimal("0.5")
        elif m <= ExactDecimal("200"):
            base_factor = ExactDecimal("1.0")
        else:
            base_factor = ExactDecimal("1.5")

    elif accuracy_class == AccuracyClassEnum.CLASS_II:
        # Class II stepped boundaries: 5000e, 20000e, 100000e
        if m <= ExactDecimal("5000"):
            base_factor = ExactDecimal("0.5")
        elif m <= ExactDecimal("20000"):
            base_factor = ExactDecimal("1.0")
        else:
            base_factor = ExactDecimal("1.5")

    elif accuracy_class == AccuracyClassEnum.CLASS_I:
        # Class I stepped boundaries: 50000e, 200000e
        if m <= ExactDecimal("50000"):
            base_factor = ExactDecimal("0.5")
        elif m <= ExactDecimal("200000"):
            base_factor = ExactDecimal("1.0")
        else:
            base_factor = ExactDecimal("1.5")

    else:
        raise ValueError(f"Unsupported accuracy class: {accuracy_class}")

    return ExactDecimal(base_factor * multiplier)


def calculate_nawi_mpe(
    load: Quantity,
    e: Quantity,
    accuracy_class: AccuracyClassEnum,
    verification_type: VerificationTypeEnum,
) -> Quantity:
    """Calculate the statutory MPE as a physical quantity in the unit of e.

    Args:
        load: Applied nominal load Quantity.
        e: Verification scale interval Quantity.
        accuracy_class: Accuracy class of instrument.
        verification_type: INITIAL or RE_VERIFICATION.

    Returns:
        MPE as a Quantity matching the unit of e.
    """
    load_in_e_unit = load.to_unit(e.unit)
    m_intervals = ExactDecimal(load_in_e_unit.value / e.value)
    mpe_factor = get_nawi_mpe_factor_in_e(m_intervals, accuracy_class, verification_type)
    mpe_mass_value = ExactDecimal(mpe_factor * e.value)
    return Quantity(mpe_mass_value, e.unit)
